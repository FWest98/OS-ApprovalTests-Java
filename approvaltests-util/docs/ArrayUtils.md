<a id="top"></a>

# ArrayUtils



<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Contents**

- [Joining Collections](#joining-collections)
- [Joining Collections with a function](#joining-collections-with-a-function)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## toArray

ArrayUtils has the ability to dynamically determine the type of your list.

<!-- snippet: toArray -->
<a id='snippet-toarray'></a>
```java
List<Comparable> comparables = new ArrayList<>();
comparables.add(null);
comparables.add(1);
comparables.add(3.1415);
comparables.add("Lars");
Comparable[] comparableArray = ArrayUtils.toArray(comparables);
```
<sup><a href='/approvaltests-util-tests/src/test/java/com/spun/util/ArrayUtilsTest.java#L43-L50' title='Snippet source file'>snippet source</a> | <a href='#snippet-toarray' title='Start of snippet'>anchor</a></sup>
<!-- endSnippet -->

This works in almost all cases with the notable exceptions of empty arrays, some common interfaces 
which will not be preferenced.
In which case you can always use the overloaded method 
<!-- snippet: toArrayWithClass -->
<a id='snippet-toarraywithclass'></a>
```java
Comparable[] array = ArrayUtils.toArray(comparables, Comparable.class);
```
<sup><a href='/approvaltests-util-tests/src/test/java/com/spun/util/ArrayUtilsTest.java#L51-L53' title='Snippet source file'>snippet source</a> | <a href='#snippet-toarraywithclass' title='Start of snippet'>anchor</a></sup>
<!-- endSnippet -->


[Back to User Guide](README.md#top)
