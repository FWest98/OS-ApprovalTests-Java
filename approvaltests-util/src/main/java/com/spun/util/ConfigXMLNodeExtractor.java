package com.spun.util;

import com.spun.util.io.XMLNodeExtractor;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

import java.util.ArrayList;
import java.util.HashMap;

public class ConfigXMLNodeExtractor implements XMLNodeExtractor
{
  public void extractProperty(Node node, HashMap<String, Object> properties)
  {
    //String name = node.getNodeName();
    if (XMLNodeExtractor.Utils.extractSingleton(node, properties))
    {
      // do nothing
    }
    else if (node.getChildNodes().getLength() > 0)
    {
      NodeList childNodes = node.getChildNodes();
      ArrayList<String> v = new ArrayList<String>();
      for (int i = 0; i < childNodes.getLength(); i++)
      {
        if (childNodes.item(i).getNodeName().equals("VALUE"))
        {
          v.add(StringUtils.loadNullableString(childNodes.item(i).getChildNodes().item(0).getNodeValue()));
        }
      }
      properties.put(node.getNodeName(), StringUtils.toArray(v));
    }
  }
}
