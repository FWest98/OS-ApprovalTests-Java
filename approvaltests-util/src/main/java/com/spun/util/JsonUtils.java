package com.spun.util;

import com.google.gson.*;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonWriter;
import org.lambda.functions.Function1;

import java.io.IOException;
import java.time.Instant;

public class JsonUtils
{
  public static String prettyPrint(String json)
  {
    if (!ObjectUtils.isClassPresent("com.google.gson.Gson"))
    {
      throw new RuntimeException(
          "Missing Gson dependency\n  Pretty print uses Gson parser.\n  You can get this from the maven repo \n  or https://github.com/google/gson");
    }
    try
    {
      Gson gson = new GsonBuilder().setPrettyPrinting().create();
      JsonElement je = JsonParser.parseString(json);
      return gson.toJson(je);
    }
    catch (JsonSyntaxException e)
    {
      return String.format("Error:%s\nJson:\n%s", e.getMessage(), json);
    }
  }
  public static String asJson(Object o)
  {
    return asJson(o, (b) -> b);
  }
  public static <T> String asJsonWithBuilder(Object o, Function1<T, T> gsonBuilder, Class<T> gsonBuilderClass)
  {
    if (!ObjectUtils.isThisInstanceOfThat(gsonBuilderClass, com.google.gson.GsonBuilder.class))
    { throw new FormattedException("Class must be of type %s", GsonBuilder.class.getName()); }
    return asJson(o, (Function1<GsonBuilder, GsonBuilder>) gsonBuilder);
  }
  public static <T> String asJson(Object o, Function1<GsonBuilder, GsonBuilder> gsonBuilder)
  {
    GsonBuilder builder = new GsonBuilder().setPrettyPrinting();
    builder = addHandlingForDateObjects(builder);
    Gson gson = gsonBuilder.call(builder).create();
    return gson.toJson(o);
  }
  private static GsonBuilder addHandlingForDateObjects(GsonBuilder builder)
  {
    builder = builder.registerTypeAdapter(Instant.class, new InstantAdapter());
    return builder;
  }
  public static class InstantAdapter extends TypeAdapter<Instant>
  {
    @Override
    public void write(JsonWriter jsonWriter, Instant instant) throws IOException
    {
      jsonWriter.value(instant.toString());
    }
    @Override
    public Instant read(JsonReader jsonReader) throws IOException
    {
      throw new IOException("Never called");
    }
  }
}
