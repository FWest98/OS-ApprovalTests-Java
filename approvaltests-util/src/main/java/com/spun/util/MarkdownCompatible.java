package com.spun.util;

public interface MarkdownCompatible
{
  String toMarkdown();
}
