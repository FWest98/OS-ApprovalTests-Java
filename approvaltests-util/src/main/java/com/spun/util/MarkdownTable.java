package com.spun.util;

import org.lambda.functions.Function1;
import org.lambda.query.Queryable;

public class MarkdownTable implements MarkdownCompatible
{
  protected String markdown;
  public static <I, O> MarkdownTable create(I[] inputs, Function1<I, O> o, String column1, String column2)
  {
    MarkdownTable table = new MarkdownTable().withColumnHeaders(column1, column2);
    for (I input : inputs)
    {
      table.addRow(input, o.call(input));
    }
    return table;
  }
  public static MarkdownTable withHeaders(String... columnNames)
  {
    MarkdownTable table = new MarkdownTable();
    return table.withColumnHeaders(columnNames);
  }
  public <I> MarkdownTable addRowsForInputs(I[] inputs, Function1<I, Object>... transfers)
  {
    for (I input : inputs)
    {
      Queryable<Object> row = Queryable.as(transfers).select(f -> f.call(input));
      row.add(0, input);
      addRow(row.toArray());
    }
    return this;
  }
  public MarkdownTable addRow(Object... columns)
  {
    markdown += printRow(columns);
    return this;
  }
  public MarkdownTable withColumnHeaders(String... headers)
  {
    markdown = printColumnHeaders(headers);
    return this;
  }
  @Override
  public String toString()
  {
    return toMarkdown();
  }
  @Override
  public String toMarkdown()
  {
    return markdown;
  }
  public static String printColumnHeaders(String... headers)
  {
    return printRow(headers) + printRow(ArrayUtils.of("---", headers.length));
  }
  public static String printRow(Object... columns)
  {
    StringBuffer b = new StringBuffer();
    b.append("|");
    for (int x = 0; x < columns.length; ++x)
    {
      b.append(String.format(" %s |", columns[x]));
    }
    b.append("\n");
    return b.toString();
  }
}
