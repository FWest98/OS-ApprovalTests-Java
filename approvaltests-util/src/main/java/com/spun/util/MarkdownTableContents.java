package com.spun.util;

public class MarkdownTableContents implements MarkdownTableElement, Resizable
{
  private String contents;
  private int    padUntil;
  public MarkdownTableContents(String contents)
  {
    this.contents = contents;
  }
  @Override
  public String toString()
  {
    return " " + StringUtils.pad(contents, padUntil) + " ";
  }
  public int getLength()
  {
    return contents.length();
  }
  public void setPadding(int length)
  {
    padUntil = length;
  }
}
