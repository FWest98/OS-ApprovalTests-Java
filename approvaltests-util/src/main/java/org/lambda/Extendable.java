package org.lambda;

public interface Extendable<T>
{
  public void setCaller(T caller);
}
