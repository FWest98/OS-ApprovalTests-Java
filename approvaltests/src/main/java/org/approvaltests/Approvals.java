package org.approvaltests;

import com.spun.util.*;
import com.spun.util.persistence.ExecutableQuery;
import com.spun.util.persistence.Loader;
import com.spun.util.persistence.SqlLoader;
import org.approvaltests.approvers.ApprovalApprover;
import org.approvaltests.approvers.FileApprover;
import org.approvaltests.core.ApprovalFailureReporter;
import org.approvaltests.core.ApprovalWriter;
import org.approvaltests.namer.ApprovalNamer;
import org.approvaltests.namer.StackTraceNamer;
import org.approvaltests.reporters.ExecutableQueryFailure;
import org.approvaltests.writers.*;
import org.lambda.actions.Action0;
import org.lambda.functions.Function1;
import org.lambda.query.Query;

import java.io.File;
import java.io.FileFilter;
import java.io.FilenameFilter;
import java.sql.ResultSet;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;

public class Approvals
{
  public static Loader<ApprovalNamer> namerCreater = new Loader<ApprovalNamer>()
  {
    public ApprovalNamer load()
    {
      return new StackTraceNamer();
    }
  };
  public static void verify(String response)
  {
    verify(new ApprovalTextWriter(response, "txt"));
  }
  public static void verify(Object o)
  {
    verify("" + o);
  }
  public static <T> void verifyAll(String label, T[] array)
  {
    verify(new ApprovalTextWriter(StringUtils.toString(label, array), "txt"));
  }
  public static <T> void verifyAll(String header, String label, T[] array)
  {
    verify(new ApprovalTextWriter(formatHeader(header) + StringUtils.toString(label, array), "txt"));
  }
  public static <T> void verifyAll(T[] values, Function1<T, String> f1)
  {
    String text = ArrayUtils.toString(values, f1);
    verify(new ApprovalTextWriter(text, "txt"));
  }
  public static <T> void verifyAll(String header, T[] values, Function1<T, String> f1)
  {
    verifyAll(header, Arrays.asList(values), f1);
  }
  public static <T> void verifyAll(String header, Iterable<T> array, Function1<T, String> f1)
  {
    String text = formatHeader(header) + ArrayUtils.toString(array, f1);
    verify(new ApprovalTextWriter(text, "txt"));
  }
  private static String formatHeader(String header)
  {
    return StringUtils.isEmpty(header) ? "" : header + "\n\n\n";
  }
  public static <T> void verifyAll(String label, Iterable<T> array)
  {
    verify(new ApprovalTextWriter(StringUtils.toString(label, array), "txt"));
  }
  public static <T> void verifyAll(String header, String label, Iterable<T> array)
  {
    verify(new ApprovalTextWriter(formatHeader(header) + StringUtils.toString(label, array), "txt"));
  }

  public static void verifyHtml(String response)
  {
    verify(new ApprovalTextWriter(response, "html"));
  }
  public static void verify(File generateFile)
  {
    verify(new FileApprovalWriter(generateFile));
  }

  public static void verify(ApprovalWriter writer, ApprovalNamer namer, ApprovalFailureReporter reporter)
  {
    verify(new FileApprover(writer, namer), reporter);
  }
  public static void verify(ApprovalWriter writer)
  {
    verify(writer, createApprovalNamer(), getReporter());
  }
  public static void verifyXml(String xml)
  {
    verify(new ApprovalXmlWriter(xml));
  }
  public static void verify(ApprovalApprover approver)
  {
    verify(approver, getReporter());
  }
  public static void verify(ApprovalApprover approver, ApprovalFailureReporter reporter)
  {
    if (!approver.approve())
    {
      approver.reportFailure(reporter);
      approver.fail();
    }
    else
    {
      approver.cleanUpAfterSuccess(reporter);
    }
  }
  public static void verify(ExecutableQuery query)
  {
    verify(new ApprovalTextWriter(query.getQuery(), "txt"), createApprovalNamer(),
        new ExecutableQueryFailure(query));
  }
  public static void verify(Map<?, ?> map)
  {
    verify(new ApprovalTextWriter(StringUtils.toString(map), "txt"));
  }
  public static void verify(ResultSet rs)
  {
    verify(new ResultSetApprovalWriter(rs));
  }
  public static <T> void verify(SqlLoader<T> loader)
  {
    verify(new SqlLoader.ExecutableWrapper<T>(loader));
  }
  public static ApprovalNamer createApprovalNamer()
  {
    return namerCreater.load();
  }

  public static void verifyEachFileInDirectory(File directory)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles());
  }
  public static void verifyEachFileInDirectory(File directory, FileFilter filter)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles(filter));
  }
  public static void verifyEachFileInDirectory(File directory, FilenameFilter filter)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles(filter));
  }
  private static void verifyEachFileAgainstMasterDirectory(File[] files) throws Error
  {
    ApprovalNamer namer = createApprovalNamer();
    String dirName = namer.getSourceFilePath() + File.separator + namer.getApprovalName() + ".Files";
    File approvedDirectory = new File(dirName);
    List<File> mismatched = new ArrayList<File>();
    for (File f : files)
    {
      if (!f.isDirectory())
      {
        try
        {
          verify(new DirectoryToDirectoryWriter(f, approvedDirectory));
        }
        catch (Throwable e)
        {
          mismatched.add(f);
        }
      }
    }
    if (!mismatched.isEmpty())
    {
      String message = "The Following Files did not match up: " + getFileNameList(mismatched);
      throw new Error(message);
    }
  }
  private static String getFileNameList(List<File> mismatched)
  {
    return Query.select(mismatched, a -> a.getName()).toString();
  }
  public static void verifyJson(String json)
  {
    verify(JsonUtils.prettyPrint(json), "json");
  }
  public static void verify(Object actual, String fileExtensionWithoutDot)
  {
    verify(new ApprovalTextWriter("" + actual, fileExtensionWithoutDot));
  }
  public static ApprovalFailureReporter getReporter()
  {
    return ReporterFactory.get();
  }
  public static void verifyAsJson(Object o)
  {
    verify(JsonUtils.asJson(o), "json");
  }
  public static void verifyException(Action0 runnableBlock)
  {
    Throwable t = ObjectUtils.captureException(runnableBlock);
    if (t == null) { throw new FormattedException("No exception thrown when running %s", runnableBlock); }
    Approvals.verify(String.format("%s: %s", t.getClass().getName(), t.getMessage()));
  }
}