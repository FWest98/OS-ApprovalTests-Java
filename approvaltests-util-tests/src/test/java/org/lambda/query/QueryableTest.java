package org.lambda.query;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.ArrayList;
import java.util.List;

import com.spun.util.ArrayUtils;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;
import org.lambda.Extendable;

class QueryableTest
{
  // begin-snippet: implementing-extendable
  public static class CustomQuery implements Extendable<List<String>>
  {
    private List<String> caller;
    @Override
    public void setCaller(List<String> caller)
    {
      this.caller = caller;
    }
    // end-snippet
    // begin-snippet: extendable-query
    public Queryable<String> findFirstWordsOnly()
    {
      return findFirstWordsOnly(caller);
    }
    // end-snippet
    // begin-snippet: custom-query
    public static Queryable<String> findFirstWordsOnly(List<String> words)
    {
      return Query.select(words, w -> {
        int i = w.indexOf(' ');
        if (i == -1)
        {
          return w;
        }
        else
        {
          return w.substring(0, i);
        }
      });
    }
    // end-snippet
  }
  @Test
  void example()
  {
    // begin-snippet: custom-query-example
    Queryable<String> list = Queryable.as("One fish", "two fish", "red fish", "blue fish");
    Queryable<String> firstWordsOnlyWithExtension = list.select(String::toUpperCase).use(CustomQuery.class)
        .findFirstWordsOnly();
    // end-snippet
    // begin-snippet: custom-query-example-static
    Queryable<String> firstWordsOnlyStatic = CustomQuery
        .findFirstWordsOnly(Query.select(list, String::toUpperCase));
    // end-snippet
    assertArrayEquals(firstWordsOnlyStatic.toArray(), firstWordsOnlyWithExtension.toArray());
    Approvals.verifyAll("firstWordsOnly", firstWordsOnlyWithExtension);
  }
  @Test
  void testDistinct()
  {
    Queryable<Integer> distinct = Queryable.as(3, 2, 1, 1, 3).distinct();
    Approvals.verifyAll("", distinct);
  }
  @Test
  void testToArray()
  {
    Character[] letters = Queryable.as('L', 'a', 'r', 's').asArray();
    Approvals.verifyAll("", letters);
  }
  @Test
  void testEmptyArrays()
  {
    String[] strings = new String[0];
    Queryable<String> as = Queryable.as(strings);
    as.add("Lars");
    String[] asArray = as.asArray();
    Approvals.verifyAll("", asArray);
  }
  @Test
  void testEmptyList()
  {
    List<String> strings = new ArrayList<>();
    Queryable<String> as = Queryable.as(strings, String.class); // Queryable.as(strings); ?
    as.add("Lars");
    String[] asArray = as.asArray();
    Approvals.verifyAll("", asArray);
  }
  @Test
  void testSuperClassCommonality()
  {
    List<Number> numbers = new ArrayList<>();
    numbers.add(1);
    numbers.add(3.1415);
    Queryable<Number> as = Queryable.as(numbers); // Queryable.as(numbers); ?
    as.add(2.4);
    Number[] asArray = as.asArray();
    Approvals.verifyAll("", asArray);
  }
  @Test
  void testInterfaceCommonality()
  {
    List<Comparable> comparables = new ArrayList<>();
    comparables.add(null);
    comparables.add(1);
    comparables.add(3.1415);
    comparables.add("Lars");
    Queryable<Comparable> queryable = Queryable.as(comparables);
    assertEquals(Comparable.class, queryable.getType());
    queryable.add(2.4);
    Comparable[] asArray = queryable.asArray();
    Approvals.verifyAll("", asArray);
  }
  @Test
  void testInterfaceCommonality2()
  {
    List<Comparable> comparables = new ArrayList<>();
    comparables.add(null);
    comparables.add(1);
    comparables.add(3.1415);
    comparables.add("Lars");
    Comparable[] itCompilesAndRuns = ArrayUtils.toArray(comparables);
    Approvals.verifyAll("", itCompilesAndRuns);
  }
}
