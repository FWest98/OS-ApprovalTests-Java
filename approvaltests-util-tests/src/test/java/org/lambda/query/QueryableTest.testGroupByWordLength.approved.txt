3 = [One, Two, Red]
4 = [Fish, Fish, Fish, Blue, Fish]
